File: backend/src/bond_monitor/domain/shared/money.py
"""
Строгие денежные/количественные типы для домена режима торговли.

Облигации торгуются в **процентах от номинала** (например, цена ``100.5%``
означает ``1005 ₽`` при номинале ``1000 ₽``), а лимитные заявки в API
T-Invest принимают цену как `Quotation` (units + nano) — структуру с
фиксированной точкой. Между этими тремя представлениями лёгкая каша
(`float` рубли vs `float` проценты vs `Quotation`) — частый источник
ошибок «купил/продал не туда».

Здесь определены `NewType`-обёртки, маркирующие смысл значения, и
явные конвертеры. Применяются агрессивно во всех новых модулях
(`trading_client`, `pending_operations`, `yield_calc`, `portfolio_reconciler`).
В существующих симуляционных модулях — точечно, по местам соприкосновения
с торговым слоем, без сплошного рефакторинга.

Примечание:
    `NewType` — лёгкая аннотация для статического анализа (`mypy`,
    `pyright`), в рантайме `Rub(1000.0)` это просто `float`. Это
    сознательный компромисс: мы хотим ловить ошибки в IDE/CI, но не
    платить за value-объекты.
"""

from __future__ import annotations

from decimal import ROUND_HALF_UP, Decimal
from typing import TYPE_CHECKING, NewType

if TYPE_CHECKING:
    # Импортируем Quotation только для type-hints, чтобы основной импорт
    # `core.money` не требовал `t_tech.invest` (модуль чистый, должен
    # импортироваться в тестах без внешних зависимостей).
    from t_tech.invest import Quotation


# Сумма в российских рублях, включая копейки. Положительная — приход,
# отрицательная — расход (как в `CashflowEvent.amount_rub`).
Rub = NewType("Rub", float)


# Цена облигации в процентах от номинала. ``100.0`` = по номиналу,
# ``99.5`` = с дисконтом, ``101.5`` = с премией. Используется в MOEX
# (`last_price`), в `BondRecord.last_price`, в `PortfolioPosition.purchase_clean_price_pct`,
# и в полях `price_pct` всех новых dataclass-ов trading-слоя.
PriceUnitPct = NewType("PriceUnitPct", float)


# Количество лотов. 1 лот = ``lot_size`` облигаций (от 1 до 100+).
Lots = NewType("Lots", int)


# Хард-лимит T-Invest API на сумму одной заявки. Заявки сверх лимита
# требуют SMS-подтверждения, которое через API не доступно (см.
# AGENTS.md «Режим торговли»). Все суммы в ``data.trading_client``
# проверяются на этот порог ДО `post_order`.
MAX_ORDER_AMOUNT_RUB: Rub = Rub(30_000_000.0)


def lot_cost_rub(
    *,
    price_pct: PriceUnitPct,
    face_value: float,
    lot_size: int,
    aci_rub: float = 0.0,
) -> Rub:
    """Стоимость одного лота облигации в рублях (грязная цена).

    Args:
        price_pct: Цена в % от номинала (например, ``100.5``).
        face_value: Номинал одной облигации в ₽ (``1000.0`` по умолчанию
            для большинства корпоративных бумаг).
        lot_size: Количество облигаций в одном лоте.
        aci_rub: НКД на одну облигацию в ₽ (если уже учтён в price_pct —
            передавать ``0.0``).

    Returns:
        Грязная цена одного лота в ₽, как ``Rub``.
    """
    clean_per_bond = price_pct / 100.0 * face_value
    dirty_per_bond = clean_per_bond + aci_rub
    return Rub(dirty_per_bond * lot_size)


def order_amount_rub(
    *,
    price_pct: PriceUnitPct,
    face_value: float,
    lot_size: int,
    lots: Lots,
    aci_rub: float = 0.0,
) -> Rub:
    """Общая стоимость заявки (lots × lot × dirty)."""
    return Rub(
        lot_cost_rub(
            price_pct=price_pct,
            face_value=face_value,
            lot_size=lot_size,
            aci_rub=aci_rub,
        )
        * lots
    )


def pct_to_quotation(price_pct: PriceUnitPct) -> Quotation:
    """Сконвертировать числовое значение в proto `Quotation` (units, nano).

    Нормализация до 9 знаков после запятой (формат `Quotation.nano`).
    """
    from t_tech.invest.utils import decimal_to_quotation

    quantized = Decimal(str(price_pct)).quantize(Decimal("0.000000001"), rounding=ROUND_HALF_UP)
    return decimal_to_quotation(quantized)


def bond_clean_price_quotation(*, price_pct: PriceUnitPct, face_value: float) -> Quotation:
    """Чистая цена одной облигации в ₽ для `PostOrder` / `GetOrderPrice`.

    В котировках MOEX и в UI цена хранится как % от номинала, а T-Invest API
    для лимитных заявок по облигациям принимает **рублёвую чистую цену за 1 бумагу**.
    """
    clean_per_bond = float(price_pct) / 100.0 * face_value
    return pct_to_quotation(PriceUnitPct(clean_per_bond))


def quotation_to_pct(q: Quotation) -> PriceUnitPct:
    """Обратная конвертация `Quotation` → ``% от номинала`` (как ``float``)."""
    from t_tech.invest.utils import quotation_to_decimal

    return PriceUnitPct(float(quotation_to_decimal(q)))


def money_value_to_rub(mv: object) -> Rub | None:
    """Сконвертировать proto `MoneyValue` в `Rub`.

    Принимает `MoneyValue` (через duck-typing: атрибуты ``units``, ``nano``,
    ``currency``). Возвращает ``None`` для нулевых значений или валют
    отличных от RUB (всё, что не RUB, в портфельной модели не учитывается —
    привязка только к RUB-счёту, см. AGENTS.md).
    """
    if mv is None:
        return None
    currency = getattr(mv, "currency", "").lower()
    if currency and currency != "rub":
        return None
    units = int(getattr(mv, "units", 0))
    nano = int(getattr(mv, "nano", 0))
    return Rub(float(units) + nano / 1_000_000_000.0)


def quotation_to_float(q: object) -> float | None:
    """Сконвертировать proto `Quotation` в обычный `float` (с дробной частью).

    Возвращает ``None`` если ``q`` пустой / нулевой по обеим компонентам
    (для опциональных полей API). Дробная часть `nano` — 9 знаков.
    """
    if q is None:
        return None
    units = int(getattr(q, "units", 0))
    nano = int(getattr(q, "nano", 0))
    if units == 0 and nano == 0:
        return None
    return float(units) + nano / 1_000_000_000.0

File: backend/internal/domain/shared/money.go
// Package shared provides strict monetary/quantity types for the trading domain.
package shared

import (
	"math"
)

// Rub is an amount in Russian rubles (including kopecks).
type Rub float64

// PriceUnitPct is a bond price as percent of face value (100.0 = par).
type PriceUnitPct float64

// Lots is the number of exchange lots.
type Lots int

// MaxOrderAmountRub is the T-Invest API hard limit per order.
const MaxOrderAmountRub Rub = 30_000_000

// Quotation mirrors proto Quotation (units + nano fixed-point).
type Quotation struct {
	Units int64
	Nano  int64
}

// MoneyValue is the duck-typed proto MoneyValue interface.
type MoneyValue interface {
	Currency() string
	Units() int64
	Nano() int64
}

// QuotationReader is the duck-typed proto Quotation interface.
type QuotationReader interface {
	Units() int64
	Nano() int64
}

// LotCostRub returns dirty price of one lot in rubles.
func LotCostRub(pricePct PriceUnitPct, faceValue float64, lotSize int, aciRub float64) Rub {
	cleanPerBond := float64(pricePct) / 100.0 * faceValue
	dirtyPerBond := cleanPerBond + aciRub
	return Rub(dirtyPerBond * float64(lotSize))
}

// OrderAmountRub returns total order cost (lots × lot dirty price).
func OrderAmountRub(pricePct PriceUnitPct, faceValue float64, lotSize int, lots Lots, aciRub float64) Rub {
	return Rub(float64(LotCostRub(pricePct, faceValue, lotSize, aciRub)) * float64(lots))
}

// PctToQuotation converts a numeric value to Quotation (9 decimal places, half-up).
func PctToQuotation(pricePct PriceUnitPct) Quotation {
	quantized := roundHalfUp9(float64(pricePct))
	units := int64(quantized)
	frac := quantized - float64(units)
	nano := int64(math.Round(frac * 1_000_000_000))
	if nano == 1_000_000_000 {
		units++
		nano = 0
	}
	return Quotation{Units: units, Nano: nano}
}

// BondCleanPricePctFromRub converts clean ruble price per bond to % of face value.
func BondCleanPricePctFromRub(cleanPriceRub, faceValue float64) PriceUnitPct {
	if faceValue <= 0 {
		panic("face_value must be positive")
	}
	return PriceUnitPct(cleanPriceRub / faceValue * 100.0)
}

// BondCleanPriceQuotation converts % of face to ruble clean price Quotation per bond.
func BondCleanPriceQuotation(pricePct PriceUnitPct, faceValue float64) Quotation {
	cleanPerBond := float64(pricePct) / 100.0 * faceValue
	return PctToQuotation(PriceUnitPct(cleanPerBond))
}

// QuotationToPct converts Quotation back to a float percent/price value.
func QuotationToPct(q Quotation) PriceUnitPct {
	return PriceUnitPct(float64(q.Units) + float64(q.Nano)/1_000_000_000.0)
}

// MoneyValueToRub converts proto MoneyValue to Rub (nil for non-RUB or nil input).
func MoneyValueToRub(mv MoneyValue) *Rub {
	if mv == nil {
		return nil
	}
	currency := mv.Currency()
	if currency != "" && currency != "rub" {
		return nil
	}
	v := Rub(float64(mv.Units()) + float64(mv.Nano())/1_000_000_000.0)
	return &v
}

// QuotationToFloat converts proto Quotation to float (nil for zero/empty).
func QuotationToFloat(q QuotationReader) *float64 {
	if q == nil {
		return nil
	}
	units := q.Units()
	nano := q.Nano()
	if units == 0 && nano == 0 {
		return nil
	}
	v := float64(units) + float64(nano)/1_000_000_000.0
	return &v
}

func roundHalfUp9(v float64) float64 {
	return math.Round(v*1_000_000_000) / 1_000_000_000
}
